File: src/main/java/red/sukun1899/service/TableService.java
package red.sukun1899.service;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import red.sukun1899.AppConfig;
import red.sukun1899.model.ReferencedTableCount;
import red.sukun1899.model.Table;
import red.sukun1899.repository.TableRepository;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * @author su-kun1899
 */
@Service
public class TableService {
  private final AppConfig appConfig;
  private final TableRepository tableRepository;

  public TableService(AppConfig appConfig, TableRepository tableRepository) {
    this.appConfig = appConfig;
    this.tableRepository = tableRepository;
  }

  @Transactional(readOnly = true)
  public List<Table> get() {
    return tableRepository.selectAll(appConfig.getSchemaName());
  }

  @Transactional(readOnly = true)
  public Table get(String tableName) {
    return tableRepository.select(appConfig.getSchemaName(), tableName);
  }

  /**
   * @return Key: tableName, Value: Parent table's count
   */
  public Map<String, Long> getParentTableCountsByTableName() {
    Map<String, ReferencedTableCount> referencedTableCountMap =
        tableRepository.selectParentTableCountsByTableName(appConfig.getSchemaName());

    return referencedTableCountMap.entrySet().stream()
        .collect(
            Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().getCount())
        );
  }
}

File: src/main/kotlin/red/sukun1899/shishamo/service/TableService.kt
package red.sukun1899.shishamo.service

import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import red.sukun1899.shishamo.model.DataDefinition
import red.sukun1899.shishamo.model.Table
import red.sukun1899.shishamo.model.TableDetails
import red.sukun1899.shishamo.repository.TableRepository

/**
 * @author su-kun1899
 */
@Service
class TableService(
        val dataSourceProperties: DataSourceProperties,
        val tableRepository: TableRepository
) {
    @Transactional(readOnly = true)
    fun get(): List<Table> = tableRepository.selectAll(dataSourceProperties.name)

    @Transactional(readOnly = true)
    fun get(tableName: String): TableDetails = tableRepository.select(dataSourceProperties.name, tableName)

    @Transactional(readOnly = true)
    fun getCreateTableStatement(table: Table): DataDefinition = tableRepository.showCreateTableStatement(table)

    /**
     * @return Key: table, Value: Parent table's references
     */
    fun getParentTableCountsByTableName(): Map<String, Long> = tableRepository
            .selectParentTableCountsByTableName(dataSourceProperties.name)
            .filter { it.value.references > 0 }
            .map { it.key to it.value.references }
            .toMap()

    /**
     * @return Key: table, Value: Child table's references
     */
    fun getChildTableCountsByTableName(): Map<String, Long> = tableRepository
            .selectChildTableCountsByTableName(dataSourceProperties.name)
            .filter { it.value.references > 0 }
            .map { it.key to it.value.references }
            .toMap()

    /**
     * @return Key: table, Value: Column references
     */
    fun getColumnCountsByTableName(): Map<String, Long> = tableRepository
            .selectColumnCountsByTableName(dataSourceProperties.name)
            .filter { it.value.references > 0 }
            .map { it.key to it.value.references }
            .toMap()
}
